src/ripple/basics/impl/PerfLogImp.h

```diff
@@ -0,0 +1,224 @@
+
+/**
+ * Implementation class for PerfLog.
+ */
+class PerfLogImp
+    : public PerfLog, Stoppable
+{
+    /**
+     * Track performance counters and currently executing tasks.
+     */
+    struct Counters
+    {
+    public:
+        /**
+         * RPC performance counters.
+         */
+        struct Rpc
+        {
+            // Counters for each time a method starts and then either finishes
+            // successfully or with an exception.
+            std::uint64_t started {0};
+            std::uint64_t finished {0};
+            std::uint64_t errored {0};
+            // Cumulative duration of all finished and errored method calls.
+            microseconds duration {0};
+            // Don't change the contents of label else labelSs will point to
+            // something undefined.
+            std::string const label;
+            std::unique_ptr<Json::StaticString> const labelSs;
+            mutable std::mutex mut;
+
+            Rpc(std::string const& labelArg)
+                : label (labelArg)
+                , labelSs (std::make_unique<Json::StaticString>(label.c_str()))
+            {}
+
+            Rpc(Rpc const& orig)
+                : started (orig.started)
+                , finished (orig.finished)
+                , duration (orig.duration)
+                , label (orig.label)
+                , labelSs (std::make_unique<Json::StaticString>(
+                    label.c_str()))
+            {}
+        };
+
+        /**
+         * Job Queue task performance counters.
+         */
+        struct Jq
+        {
+            // Counters for
+            std::uint64_t queued {0};
+            std::uint64_t started {0};
+            std::uint64_t finished {0};
+            // Cumulative duration of all jobs' queued and running times.
+            microseconds queuedDuration {0};
+            microseconds runningDuration {0};
+            // Don't change the contents of label else labelSs will point to
+            // something undefined.
+            std::string const label;
+            std::unique_ptr<Json::StaticString> const labelSs;
+            mutable std::mutex mut;
+
+            Jq(std::string const& labelArg)
+                : label (labelArg)
+                , labelSs (std::make_unique<Json::StaticString>(label.c_str()))
+            {}
+
+            Jq(Jq const& orig)
+            : queued (orig.queued)
+            , started (orig.started)
+            , finished (orig.finished)
+            , queuedDuration (orig.queuedDuration)
+            , runningDuration (orig.runningDuration)
+            , label (orig.label)
+            , labelSs (std::make_unique<Json::StaticString>(
+                    label.c_str()))
+            {}
+        };
+
+        std::unordered_map<std::string, Rpc> rpc_;
+        std::unordered_map<std::underlying_type_t<JobType>, Jq> jq_;
+        std::vector<std::pair<JobType, steady_time_point>> jobs_;
+        int workers_ {0};
+        mutable std::mutex jobsMutex_;
```

Comment:
The presence of `jobsMutex_`suggests you've done some solid thinking about synchronization.  And a quick audit suggests `jobsMutex_` is being used in all the right places.  However what that leaves open is whether `rpc_` and `jq_` need to be protected by a mutex as well.  I'm not saying that they necessarily do.  But, if they don't, there should at least be a comment justifying why they don't need synchronization.

And, naturally, if they do need synchronization, please add it.  Thanks.

Reply:
fixed for real